src/ripper/ripcddialog.cpp

```diff
@@ -293,20 +291,12 @@ void RipCDDialog::DeviceSelected(int device_index) {
   }
 
   connect(cdda_device_.get(), SIGNAL(DiscChanged()), SLOT(DiscChanged()));
-
-  // get SongLoader from device and connect signals
-  loader_ = cdda_device_->loader();
-  Q_ASSERT(loader_);
-
-  connect(loader_, SIGNAL(SongsDurationLoaded(SongList)),
+  connect(cdda_device_.get(), SIGNAL(SongsDiscovered(SongList)),
           SLOT(BuildTrackListTable(SongList)));
-  connect(loader_, SIGNAL(SongsMetadataLoaded(SongList)),
-          SLOT(UpdateTrackListTable(SongList)));
-  connect(loader_, SIGNAL(SongsMetadataLoaded(SongList)),
-          SLOT(AddAlbumMetadataFromMusicBrainz(SongList)));
+  connect(cdda_device_.get(), SIGNAL(SongsDiscovered(SongList)),
+          SLOT(SetAlbumMetadata(SongList)));
+  cdda_device_->LoadSongs();
```

Comment:
If I'm reading things correctly, the purpose of this is to re-emit the SongsDiscovered signal. Could this have side effects on other users?

Reply:
Maybe we could keep the song cache in the CddaDevice so that it's always up to date with the model. Then we could just fetch those when this initializes.

Reply:
This only re-emits the SongsDiscovered signal if songs were previously already loaded, however, it can also start a genuine loading process if the device wasn't yet opened/loaded. It has side effects on other consumers of the SongsDiscovered signal insofar as there will be a re-emission of that signal. However, the same signal will also happen if the disc was changed and consumers are therefore expected to (and do) treat this as an update to the song list in a reasonable way.


I was thinking of containing the song cache in CddaDevice, but that has the problem that CddaDevice cannot currently detect disc changes, which require an emptying of the cache, because CddaDevice is now the only instance that has access to the cdio object. The whole split between CddaSongLoader and CddaDevice is a bit problematic as they really heavily depend on each other to function properly but I haven't yet found a good solution to that.